src/rust/src/task/executor.rs

```diff
@@ -0,0 +1,120 @@
+use super::{Task, TaskId};
+use alloc::{collections::BTreeMap, sync::Arc, task::Wake};
+use crossbeam_queue::ArrayQueue;
+use core::task::{Context, Poll, Waker};
+
+pub struct Executor {
+    tasks: BTreeMap<TaskId, Task>,
+    task_queue: Arc<ArrayQueue<TaskId>>,
+    waker_cache: BTreeMap<TaskId, Waker>,
+}
+
+// Struct representing the executor which manages tasks and their execution.
+impl Executor {
+    // Constructor for Executor. Initializes the tasks, task_queue and waker_cache.
+    pub fn new() -> Self {
+        Executor {
+            tasks: BTreeMap::new(),
+            task_queue: Arc::new(ArrayQueue::new(100)),
+            waker_cache: BTreeMap::new(),
+        }
+    }
+
+    // Spawns a new task. 
+    // Inserts the task into tasks map and pushes the task_id into the task_queue.
+    pub fn spawn(&mut self, task: Task) {
+        let task_id = task.id;
+        if self.tasks.insert(task.id, task).is_some() {
+            panic!("task with same ID already in tasks");
+        }
+        self.task_queue.push(task_id).expect("queue full");
+    }
+
+    // Runs the executor, which continuously executes 
+    // tasks if they're ready, and sleeps if idle.
+    pub fn run(&mut self) -> ! {
+        loop {
+            self.run_ready_tasks();
+            self.sleep_if_idle();
+        }
+    }
+
+    // Runs tasks that are ready. 
+    // If a task is finished, it's removed from the tasks 
+    // and its waker is removed from the waker_cache.
+    fn run_ready_tasks(&mut self) {
+        // destructure `self` to avoid borrow checker errors
+        let Self {
+            tasks,
+            task_queue,
+            waker_cache,
+        } = self;
+
+        while let Some(task_id) = task_queue.pop() {
+            let task = match tasks.get_mut(&task_id) {
+                Some(task) => task,
+                None => continue, // task no longer exists
+            };
+            let waker = waker_cache
+                .entry(task_id)
+                .or_insert_with(|| TaskWaker::new(task_id, task_queue.clone()));
+            let mut context = Context::from_waker(waker);
+            match task.poll(&mut context) {
+                Poll::Ready(()) => {
+                    // task done -> remove it and its cached waker
+                    tasks.remove(&task_id);
+                    waker_cache.remove(&task_id);
+                }
+                Poll::Pending => {}
+            }
+        }
+    }
+
+    // Puts the executor into sleep state if no tasks are in the queue. 
+    // Otherwise, enables the interrupts.
+    fn sleep_if_idle(&self) {
+        use x86_64::instructions::interrupts::{self, enable_and_hlt};
+
```

Comment:
Check out `include/nautilus/cpu_state.h`. In particular, using `sti` and `cli` directly (which is what the `x86_64` crate is doing) might be dangerous because the scheduler can preempt our executor at any time.
```C
// Do not directly use these functions or sti/cli unless you know
// what you are doing...
// Instead, use irq_disable_save and a matching irq_enable_restore
#define enable_irqs() sti()
#define disable_irqs() cli()
```
Consider using the `irq_disable_save` and `irq_enable_restore` functions here. They are `static inline` so you may have to wrap them in `glue.c` to generate a binding.

Reply:
`idle.c` might be a better place to look.